### ISSUE TITLE
delete strategy

### ISSUE THREAD

**Issue by numiralofe** [miniclip/gonsul#9]
Hi all,

**Problem :** In some situations (my case) it can be useful run multiple gonsul intances with different configurations in order to sync multiple git repos to the same consul k/v path.

(yes i know its dirty :( but sometimes the "hammer" solution is _the only_ solution )

**Current Status:**  with the current option's that only allow or don't allow Deletes, if we run multiple gonsul instances pointing different repos to the same consul k/v path, they will delete each other and the previous sync process is lost or the sync process is stopped if we do not allow deletions.

```
[DEBUG] [Oct 24 18:29:28.723] Finished data retrieve from GIT
[DEBUG] [Oct 24 18:29:28.723] Starting data import to Consul
[ERROR] [Oct 24 18:29:28.761] We're stopping as there are deletes and Gonsul is running without delete permission
[ERROR] [Oct 24 18:29:28.761] Below is all the Consul KV paths that would be deleted
```

Without making major changes to the current code base the **possible solutions** that i can think of are: 

**1 )** remove the allowDelete flag and replace it with a policy flag called `deletePolicy` that would work like explained bellow:

### `--deleteStrategy`
> `require:` **no**  
> `default:` **NOTALLOW**  

- **`NOTALLOW`** this would be the default mode and would keep consistency to what currently is the default behaviour, to not allow delete operations.
- **`ALLOW`** In this mode deletes are allowed and the behaviour will stay as its currently working it the allowDelete flag set to true.
- **`SKIP`** In this mode, Gonsul will skip the deletion operations and just sync the files to consul k/v without any pre-check of files to be deleted.

**2 )** keep the allowDelete flag but change it from bool to string and add a 3rd option:

### `--allowDelete`
> `require:` **no**  
> `default:` **false**  

- **`false`** this would be the default mode and would keep consistency to what currently is the default behaviour, to not allow delete operations.
- **`true`** In this mode deletes are allowed and the behaviour will stay as it currently is.
- **`skip`** In this mode, Gonsul will skip the deletion operations and just sync the files to consul k/v without perform any deletions.

I know that both cases can lead to inconsistencies at the consul k/v since removed files from the repo will stay in the consul path because we are skipping deletions, or some files can be over written, but "the world is far from perfect" :) and for some situations _we need what we need_.

**Final Thoughts and Questions:**

I would like to continue using gonsul base code and not fork to a custom solution because of all the obvious advantages, as so and before i make any decision or start to do any work on this i want to ask:

- Are you guys keen to merge on master branch any of the solutions that i am mentioning ? 
- If yes, which one ? 
- if not, on your view any other 3rd way that might solve the problem more elegantly ?

Or its a clear "no go" for any of the solutions that i mention since this would break consistency on the replication mechanism and in spite that it might require more changes in the code base it would suit better the project direction to implement this the proper way ?

Thanks,
Nuno

**Comment by eduardev** [comment]
Hi @numiralofe 

The delete flag was design to fix a specific problem. Now the problem you are explaining seems pretty reasonable to me. I mean, Gonsul should make as less as possible assumptions about the use cases, and give options for it's users as long as it keeps focus on the global goal: a nice tool to migrate from git to consul.

Just to be clear, `skip` would not perform any deletes, but it would carry on working as normal

The proposed solution also seems globally fine. I guess I would prefer option 2 as it delivers less (if any) breaking change with previous versions. Even though we move from bool to a string, as a cmd line tool `gonsul --allowDelete true` would work as usual.

Would you be able to submit the PR of the changes?

**Comment by numiralofe** [comment]
Hi @eduardev .

Thanks for the reply, and going back to your questions:

>> Just to be clear, skip would not perform any deletes, but it would carry on working as normal

exactly.

>>as a cmd line tool gonsul --allowDelete true would work as usual 

exactly, the default remains to `false` and both `true` or `false` options are keeping its previous functionality.

>> I guess I would prefer option 2 as it delivers less (if any) breaking change with previous versions.

Yes i agree with you, if possible avoiding breaking changes is always a better approach :)

>> Would you be able to submit the PR of the changes?

Just submitted now the pull request:  added delete strategy changes #10

Thanks,
Nuno

**Comment by eduardev** [comment]
fixed in #10
